crates/quickjs-wasm-rs/src/js_binding/context.rs

```diff
@@ -310,6 +324,56 @@ where
     }
 }
 
+fn decode_utf8_buffer_to_js_string() -> impl FnMut(&Context, &Value, &[Value]) -> Result<Value> {
+    move |ctx: &Context, _this: &Value, args: &[Value]| {
+        if args.len() != 4 {
+            return Err(anyhow!("Expecting 4 arguments, received {}", args.len()));
+        }
+
+        let buffer = args[0].as_bytes()?;
+        let byte_offset = {
+            let byte_offset_val = &args[1];
+            if !byte_offset_val.is_repr_as_i32() {
+                return Err(anyhow!("byte_offset must be an u32"));
+            }
+            byte_offset_val.as_u32_unchecked()
+        }
+        .try_into()?;
+        let byte_length: usize = {
+            let byte_length_val = &args[2];
+            if !byte_length_val.is_repr_as_i32() {
+                return Err(anyhow!("byte_length must be an u32"));
+            }
+            byte_length_val.as_u32_unchecked()
+        }
+        .try_into()?;
+        let fatal = args[3].as_bool()?;
+
+        let view = buffer
+            .get(byte_offset..(byte_offset + byte_length))
+            .ok_or(anyhow!(
+                "Provided offset and length is not valid for provided buffer"
+            ))?;
+        let str = if fatal {
+            Cow::from(str::from_utf8(view).map_err(|_| anyhow!("The encoded data was not valid"))?)
+        } else {
+            String::from_utf8_lossy(view)
+        };
```

Comment:
I'm a bit confused here, the spec states:

> If the property is true then a decoder will throw a [TypeError](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypeError) if it encounters malformed data while decoding. If false the decoder will substitute the invalid data with the replacement character U+FFFD (�). The value of the property is set in the [TextDecoder() constructor](https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder/TextDecoder).

According to this description, shouldn't we be doing something like:

```rust
let err = str::from_utf8(view).map_err(...);

if result.is_ok() {
   // return 
}

if fatal {
   // ... type error
} else {
   // ... substitute the view with the right character
}
```

Reply:
You're right, my bad.